core/integration/odc/plugin.go

```diff
@@ -248,44 +250,97 @@ func (p *Plugin) queryPartitionStatus() {
 	}
 
 	p.cachedStatusMu.Lock()
-	// env state change detection
+
+	// state change detection
 	if p.cachedStatus != nil && p.cachedStatus.Status == odc.ReplyStatus_SUCCESS {
 		for id, partitionInfo := range response.Partitions {
-			if existingPartition, ok := p.cachedStatus.Partitions[id]; ok && (existingPartition.State != partitionInfo.State) {
-				log.WithField("level", infologger.IL_Support).
-					WithField("partition", id.String()).
-					WithField("oldState", existingPartition.State).
-					WithField("state", partitionInfo.State).
-					Info("ODC Partition state changed")
-
-				payload := map[string]interface{}{
-					"oldState":         existingPartition.State,
-					"newState":         partitionInfo.State,
-					"partitionId":      partitionInfo.PartitionId.String(),
-					"ddsSessionId":     partitionInfo.DdsSessionId,
-					"ddsSessionStatus": partitionInfo.DdsSessionStatus,
-				}
-				payloadJson, _ := json.Marshal(payload)
-
-				the.EventWriterWithTopic(TOPIC).WriteEvent(&pb.Ev_IntegratedServiceEvent{
-					Name:          "odc.partitionStateChanged",
-					EnvironmentId: id.String(),
-					Payload:       string(payloadJson[:]),
-				})
-
-				envMan := environment.ManagerInstance()
-				if envMan != nil {
-					go envMan.NotifyIntegratedServiceEvent(&event.OdcPartitionStateChangeEvent{
-						IntegratedServiceEventBase: common_event.IntegratedServiceEventBase{ServiceName: "ODC"},
-						EnvironmentId:              id,
-						State:                      partitionInfo.State,
+
+			// do we have the given partition on record already? if not, no state change detection is possible
+			if existingPartition, ok := p.cachedStatus.Partitions[id]; ok {
+
+				// detection of device state change + event publication
+				for deviceId, device := range partitionInfo.Devices {
+					oldDevice, hasDevice := existingPartition.Devices[deviceId]
+
+					oldState := "" // we presume the task didn't exist before
+
+					// if a device with this ID is already known to us from before
+					if hasDevice {
+						// if device state has changed
+						if oldDevice.State != device.State {
+							// if the state has changed, we take note of the previous state
+							oldState = oldDevice.State
+						} else {
+							// if the state hasn't changed, we bail
+							continue
+						}
+					}
+
+					device.EcsState = fairmq.ToEcsState(device.State, oldState)
+
+					// since the odc-state of the task has changed, we must publish the event
+					payload := map[string]interface{}{
```

Comment:
what is this payload? where is it defined? It seems to me that just in the middle of some file you are using "random" json payload. If there is a mistake it would be nightmare to figure out where is this coming from. Isn't it possible to create some payload struct with given members that will be translated correctly instead of using string map?

Reply:
It's very much not random and its contents are purposefully chosen. It's defined here. Yes it's possible to turn it into a payload struct :+1: